specification.md

```diff
@@ -3880,6 +3903,286 @@ execution of the state, runtimes should not create a new instance of this workfl
 wait for the defined event or events again. In these cases only the states actions should be retried
 and the received event information used for all of the issued retries.
 
+### Workflow Compensation
+
+Compensation deals with undoing or reversing the work of one or more states which have 
+already successfully completed. For example, let's say that we have charged a customer $100 for an item 
+purchase. If later the customer decides to cancel this purchase we need to undo this. One way of 
+doing that is to credit the customer $100.
+
+It's important to understand that compensation with workflows is not the same as for example rolling back
+a transaction (a strict undo). Compensating a workflow state which has successfully completed 
+might involve multiple logical steps and thus is part of the overall business logic that must be 
+defined within the workflow itself. To explain this let's use our previous example and say that when our
+customer made the item purchase, our workflow has sent her/him a confirmation email. In the case we need to 
+compensate this purchase, we cannot just "undo" the confirmation email sent. Instead, we want to 
+send a second email to the customer which includes purchase cancellation information.
+
+Compensation in Serverless Workflow must be explicitly defined by the workflow control flow logic.
+It cannot be dynamically triggered by initial workflow data, event payloads, results of service invocations, or 
+errors.
+
+#### Defining Compensation
+
+Each workflow state can define how it should be compensated via its `compensatedBy` property.
+This property references another workflow state (by it's unique name) which is responsible for the actual compensation.
```

Comment:
Good point. right now you could do that by referencing a SubFlow state in your states compensatedBy property. 
Do you think we should enhance this by also allowing this property to reference a workflow id, like in workflowId property?

Reply:
That's a good point.  Personally, I think we should encourage people to keep their main workflow and compensation logic nearby each other, because they're very closely related / tightly-coupled usually.  

My take is that we should call out in the doc that as motivation:
a) workflow & compensation logic are closely related, so we feel that users should be encouraged to maintain them together in the same workflow
b) ...but....if, compensation logic is too complex or needs to be re-usable, we support SubWorkflow state as a compensation, but we don't support directly referencing a subWorkflow in the 'compensatedBy' attribute.

Reply:
SubFlow state has the usedForCompensation property - https://github.com/serverlessworkflow/specification/blob/master/specification.md#subflow-state. 
So users can chose to use it if they like. I think I agree with your statement but maybe we should add that in a new "best practices" doc/section rather than the main spec doc. wdyt